cmd/gcp/secret_manager.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2024 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
+)
+
+// Signer implements crypto.Signer using Google Cloud Secret Manager.
+type Signer struct {
+	publicKey  crypto.PublicKey
+	privateKey crypto.PrivateKey
+}
+
+// Public returns the public key stored in the Signer object.
+func (s *Signer) Public() crypto.PublicKey {
+	return s.publicKey
+}
+
+// Sign signs digest with the private key stored in Google Cloud Secret Manager.
+// Only crypto.SHA256 and ECDSA are supported.
+func (s *Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	// Verify hash function and digest bytes length.
+	if opts == nil || opts.HashFunc() != crypto.SHA256 {
+		return nil, fmt.Errorf("unsupported hash func: %v", opts.HashFunc())
```

Comment:
will this `panic` if `opts == nil` or is the implementation of `HashFunc` safe against that?

Reply:
The `opts.HashFunc()`will panic if `opts` is `nil`.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x12cd1cf]

goroutine 597 [running]:
main.(*ECDSAWithSHA256Signer).Sign(0xc0009186a0, {0x1a15640, 0xc0000b23c0}, {0xc0007102e0, 0x20, 0x20}, {0x0, 0x0})
        /home/rogerng_google_com/static-ct/cmd/gcp/secret_manager.go:47 +0x4f
github.com/transparency-dev/static-ct.buildCp({0x1a2ec48, 0xc0009186a0}, 0x472a49?, 0xc000850b20?, {0xc0008eee70?, 0xc000850d80?, 0xc000850af0?})
        /home/rogerng_google_com/static-ct/serialize.go:95 +0x1ab
github.com/transparency-dev/static-ct.(*CpSigner).Sign(0xc0009035c0, {0xc00091e900?, 0xc000850d80?, 0xc000850bb0?})
        /home/rogerng_google_com/static-ct/serialize.go:145 +0x205
golang.org/x/mod/sumdb/note.Sign(0xc0007a1e20, {0xc000850ea0, 0x1, 0xc000710280?})
        /home/rogerng_google_com/go/pkg/mod/golang.org/x/mod@v0.21.0/sumdb/note/note.go:636 +0x344
main.newGCPStorage.WithCheckpointSignerVerifier.func1.1(0x6, {0xc000710280, 0x20, 0x20})
        /home/rogerng_google_com/go/pkg/mod/github.com/transparency-dev/trillian-tessera@v0.0.0-20240911103714-727b99dc32a7/log.go:93 +0x127
github.com/transparency-dev/trillian-tessera/storage/gcp.(*Storage).integrate.func3.2()
        /home/rogerng_google_com/go/pkg/mod/github.com/transparency-dev/trillian-tessera@v0.0.0-20240911103714-727b99dc32a7/storage/gcp/gcp.go:295 +0x176
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /home/rogerng_google_com/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:78 +0x50
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 585
        /home/rogerng_google_com/go/pkg/mod/golang.org/x/sync@v0.8.0/errgroup/errgroup.go:75 +0x96
exit status 2
```

Reply:
Ah, I understand now. You mean the err message.

Reply:
Fixed, thanks!